### PR TITLE
Minimal MNIST workload

### PR DESCRIPTION
**Pull request by znado**
A rough sketch of an implementation of our current API as outlined in the rules draft, using jax/numpy.

### CODE REVIEW

File: spec.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+  pass
+
+
+# Training algorithm track submission functions, to be filled in by the
+# submitter.
+
+
+def init_optimizer_state(
+    workload: Workload,
+    param_shapes: ParameterShapeTree,
```
Comment (mikerabbat) [inline spec.py:244]: Do we need `param_shapes` as an argument given we already take `workload`?

File: submission_runner.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  eval_results = []
+  global_step = 0
+  eval_now = False
+
+  while (is_time_remaining and not goal_reached):
+    step_rng = jax.random.fold_in(rng, global_step)
+    data_select_rng, preprocess_rng, update_rng, eval_rng = jax.random.split(
+        step_rng, 4)
+    start_time = time.time()
+    selected_train_input_batch, selected_train_label_batch = data_selection(
```
Comment (mikerabbat) [inline submission_runner.py:69]: Not a Jax user, so sorry if this is a silly question.
Could this same code run on a multi-gpu system, or is this specific to a single GPU? (Or is this currently CPU-only?)

**Review by mikerabbat — COMMENTED** [review]
This is a great start! I think it covers almost all of the main points. I've added minor comments inline and I have three high-level comments.

The main thing that seems to be missing is a way for submissions to specify the hyperparameter search space. Would it make sense for submissions to provide one more function that accepts a source of randomness and outputs a sample from the tuning search space?

The way PRNG seeds are handled here is pretty specific to Jax. Different frameworks will probably need to do this differently. Is there any utility to making the harness code as framework-agnostic as possible? Is there any concern that framework-specific differences in how PNRGs are handled can become a non-trivial source of difference between implementations?

We've been discussing the benchmark in terms of a system that has 8 GPUs. Will submissions have control over how those GPUs are used? For some workloads (e.g., a ResNet50/ImageNet-like one) it may be useful to do data-parallel training using all 8 GPUs. For others (smaller ones like CIFAR-10), that may be overkill and a submission may want to try something different. The current API doesn't seem to expose this flexibility.

File: spec.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+        ParameterTree,
+        LossType,
+        Hyperparamters,
+        int,
+        RandomState
+    ],
+    Tuple[Tensor, Tensor]]
+
+
+class Workload(metaclass=abc.ABCMeta):
```
[the remark this answers is not in the document]
Comment (znado) [reply spec.py:110]: after our chat last week, I added a `get_batch_size` function to the submission to provide workload-specific batch sizes.

File: spec.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+      self,
+      raw_input_batch: Tensor,
+      train_mean: Tensor,
+      train_stddev: Tensor) -> Tensor:
+    """return preprocessed_input_batch"""
+
+  # InitModelFn = Callable[
+  #     Tuple[ParameterShapeTree, RandomState], ParameterTree]
+  @abc.abstractmethod
+  def init_model_fn(
```
[the remark this answers is not in the document]
Comment (znado) [reply spec.py:180]: Good catch, some of these were written before I refactored functions into a class, removed.

File: spec.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+        int,
+        RandomState
+    ],
+    Tuple[Tensor, Tensor]]
+
+
+class Workload(metaclass=abc.ABCMeta):
+
+  @abc.abstractmethod
+  def has_reached_goal(self, eval_result: float) -> bool:
```
[the remark this answers is not in the document]
Comment (znado) [reply spec.py:113]: For now I think we were going to use eval performance, but absolutely record other metrics like training performance.

File: spec.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+      loss_type: LossType) -> Tensor:
+    if loss_type == LossType.SOFTMAX_CROSS_ENTROPY:
+      return jax.nn.softmax(logits_batch, axis=-1)
+    if loss_type == LossType.SIGMOID_CROSS_ENTROPY:
+      return jax.nn.sigmoid(logits_batch)
+    if loss_type == LossType.MEAN_SQUARED_ERROR:
+      return logits_batch
+
+  # LossFn = Callable[Tuple[Tensor, Tensor], Tensor]
+  # Does NOT apply regularization, which is left to the submitter to do in
```
[the remark this answers is not in the document]
Comment (znado) [reply spec.py:215]: As we discussed in last weeks meeting, we don't apply any regularization in the workload functions and leave that up to the submitters.

File: spec.py (hunk 1 above)
Comment (znado) [reply spec.py:244]: Removed.

File: submission_runner.py (hunk 2 above)
Comment (znado) [reply submission_runner.py:93]: For Jax this should work on CPU/GPU/TPU!

**Review by georgedahl — APPROVED** [review]

**Pull request merged**